File: Stack/20. Valid Parentheses/Solution.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        symbols = {'(':')', '{':'}', '[':']'}
        stack = []
        
        for c in s:
            if c in symbols:
                stack.append(c)
            else:
                if not stack or c!=symbols[stack.pop()]: return False
        
        return len(stack)==0

File: Stack/20. Valid Parentheses/Solution.java
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> map = new HashMap<>();
        map.put('{', '}');
        map.put('(', ')');
        map.put('[', ']');

        for (char c : s.toCharArray()) {
            if (map.containsKey(c)) {
                stack.push(c);
            } else if (stack.isEmpty() || c != map.get(stack.pop())) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}